### PR TITLE
fix(derive): restore support for inline generic type constraints

### PR DESCRIPTION
**Pull request by aakoshh**
## Description

Fixes the `#[NetworkBehaviour]` macro to support generic constraints on behaviours without a where clause, which was the case before v0.51.

## Notes & open questions

We encountered the following problem when upgrading from v0.49 to v0.53:

We have a [behaviour with generic type constraint](https://github.com/consensus-shipyard/ipc-ipld-resolver/blob/4888d016b639ad6519c17cd3d8ff3ea8cda94122/src/behaviour/content.rs#L81-L82) inherited from a [wrapped behaviour](https://github.com/ipfs-rust/libp2p-bitswap/blob/7dd9cececda3e4a8f6e14c200a4b457159d8db33/src/behaviour.rs#L106-L107), which we then try to bundle up with all the behaviours in our application and use `derive` on the bunch:

```rust
#[derive(NetworkBehaviour)]
pub struct Behaviour<P: StoreParams> {
...
    content: content::Behaviour<P>,
}
...
mod content {
  pub struct Behaviour<P: StoreParams> {
      inner: Bitswap<P>,
  ...
}
```

After the update, the `#[NetworkBehaviour]` macro produced the following error:

```console
error[E0277]: the trait bound `P: StoreParams` is not satisfied
    |
439 |     #[derive(NetworkBehaviour)]
    |              ^^^^^^^^^^^^^^^^ the trait `StoreParams` is not implemented for `P`
    |
```

With `cargo expand` we were able to see that the reason was the missing type constraints:

```rust
    #[doc = "`NetworkBehaviour::ToSwarm` produced by Behaviour."]
    pub enum BehaviourEvent<P>
    where
...
        content::Behaviour<P>: ::libp2p::swarm::derive_prelude::NetworkBehaviour,
    {
...
        Content(
            <content::Behaviour<P> as ::libp2p::swarm::derive_prelude::NetworkBehaviour>::ToSwarm,
        ),
    }
```

Comparing that to the output generated by v0.49:

```rust
    pub enum BehaviourEvent<P: StoreParams> {
...
        Content(
            <content::Behaviour<P> as ::libp2p::swarm::derive_prelude::NetworkBehaviour>::OutEvent,
        ),
    }
```

Note that `BehaviourEvent<P: StoreParams>` allows `content::Behaviour<P> as ::libp2p::swarm::derive_prelude::NetworkBehaviour`, but it's missing in the newer version. 

During testing I noticed that this can be solved by rearranging the generics like so:

```rust
#[derive(NetworkBehaviour)]
pub struct Behaviour<P>
where 
  P: StoreParams 
{
    ...
    content: content::Behaviour<P>,
}
```

I'm not sure if this was a deliberate change and this intended to be the only way to declare generic constraints on behaviours, but with the restoring of the [original technique](https://github.com/libp2p/rust-libp2p/blob/v0.50.1/swarm-derive/src/lib.rs#L134) the code compiles as it used to be, and no other tests seem to complain, which to me seems better since both definitions are valid Rust code.

### Test

Without the fix, the test fails to compile:

```console
❯ cargo  +nightly -Zlints -q test -p libp2p-swarm --features=macros --test swarm_derive
error[E0277]: the trait bound `A: Mark` is not satisfied
   --> swarm/tests/swarm_derive.rs:439:14
    |
439 |     #[derive(NetworkBehaviour)]
    |              ^^^^^^^^^^^^^^^^ the trait `Mark` is not implemented for `A`
    |
note: required by a bound in `with_generics_constrained::Bar`
   --> swarm/tests/swarm_derive.rs:382:19
    |
382 |     struct Bar<A: Mark> {
    |                   ^^^^ required by this bound in `Bar`
    = note: this error originates in the derive macro `NetworkBehaviour` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `libp2p-swarm` (test "swarm_derive") due to previous error
```

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/tests/swarm_derive.rs [hunk 1]
```diff
@@ -371,6 +371,98 @@ fn with_generics_mixed() {
...
+    struct Bar<A: Mark> {
+        a: A,
+    }
+
+    enum BarEvent {}
+
+    impl<A: Mark + 'static> NetworkBehaviour for Bar<A> {
+        // Doesn't matter for the test, just using something to satisfy the compiler.
+        type ConnectionHandler =
+            <libp2p_identify::Behaviour as NetworkBehaviour>::ConnectionHandler;
```
Comment (thomaseizinger) [inline swarm/tests/swarm_derive.rs]: Use `dummy::ConnectionHandler` instead :)

File: swarm-derive/Cargo.toml [hunk 2]
```diff
@@ -3,7 +3,7 @@ name = "libp2p-swarm-derive"
 edition = "2021"
 rust-version = { workspace = true }
 description = "Procedural macros of libp2p-swarm"
-version = "0.34.1"
+version = "0.34.2"
```
Comment (thomaseizinger) [inline swarm-derive/Cargo.toml:6]: You'll also need to bump the version in the root manifest!

**Review by thomaseizinger — APPROVED** [review]
Given the simplicity of the fix, I am okay with this :)

File: swarm/tests/swarm_derive.rs (hunk 1 above)
Comment (aakoshh) [reply swarm/tests/swarm_derive.rs]: Thanks!

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**